Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,3028 @@
+            // small, this is unlikely to happen often.
+            bool WrappingStatusChanged = false;
+            {
+                int ix = -1;
+                for (const CurveSegment& curve : cable.m_curveSegments) {
+                    WrappingStatusChanged |=
+                        prevWrappingStatus.at(++ix) !=
+                        curve.getDataInst(sCopy).wrappingStatus;
+                }
+            }
+            if (WrappingStatusChanged) {
+                os << "Wrapping status changed: Stopping test\n";
+                break;
+            }
+
+            const Real predictionError =
+                (predictedPathError - data.pathError).norm();
+
+            bool passedTest = std::abs(predictionError / perturbation) <= bound;
+            if (!passedTest) {
+                os << "FAILED perturbation test for correction = "
+                   << data.pathCorrection << "\n";
+                os << "    Got      : " << data.pathError << "\n";
+                os << "    Predicted: " << predictedPathError << "\n";
+                os << "    Err      : "
+                   << (predictedPathError - data.pathError) / perturbation
+                   << "\n";
+                os << "    Err norm : "
+                   << (predictedPathError - data.pathError).normInf() /
+                          perturbation
+                   << "\n";
+            } else {
+                os << "PASSED perturbation test for correction = "
+                   << data.pathCorrection << "\n";
+                os << " ( error = "
+                   << (predictedPathError - data.pathError).normInf() /
+                          perturbation
+                   << " )\n";
+            }
+            success &= passedTest;
+        }
+    }
+    return success;
+}
+
+//==============================================================================
+//                          CableSubsystem::Impl
+//==============================================================================
+
+const CableSpan::Impl& CableSubsystem::Impl::getCableImpl(
+    CableSpanIndex index) const
+{
+    return getCable(index).getImpl();
+}
+
+int CableSubsystem::Impl::realizeSubsystemTopologyImpl(State& state) const
+{
+    // Briefly allow writing into the Topology cache. After this the
+    // Topology cache is const.
+    Impl* mutableThis = const_cast<Impl*>(this);
+
+    CableSpanData::Instance dataInst{};
+    CacheEntryIndex indexDataInst = allocateCacheEntry(
+        state,
+        Stage::Instance,
+        Stage::Infinity,
+        new Value<CableSpanData::Instance>(dataInst));
+
+    for (CableSpanIndex ix(0); ix < cables.size(); ++ix) {
+        CableSpan& path = mutableThis->updCable(ix);
+        path.updImpl().realizeTopology(state, indexDataInst);
+    }
+
+    return 0;
+}
+
+const CableSpan& CableSubsystem::Impl::getCable(CableSpanIndex index) const
+{
+    SimTK_ERRCHK1_ALWAYS(
```

Comment:
...and range-check the index?

Reply:
it is range checked by SimTK::Array_ (in Debug build)